README.md
About the plugin
================

Changes and versions
--------------------

**Version 1.4, released Jul 18 2013**

In this version one bug in folder context menu has been resolved. Now this menu is displayed 
properly with all items that dropbox provides.

**Version 1.3, released Jul 7 2013**

Since this version it's not possible to generate links to albums, modern Dropbox versions
could generate "album" for literally any folder inside your Dropbox account, so use 
context menu option "Share folder and copy link".

Also starting with version 1.3 plugin doesn't require web browser to generate "shared" link,
it's copied to clipboard just after you select "Share folder/file and copy link" option
in the context menu.

Important notice
----------------

This plugin does **not** include Dropbox client itself, it's just a plugin for Dolphin File Manager.

Also pay attention that plugin usually is working correctly only with the recent versions of dropbox
applications, so very old version won't work properly.

You have to install Dropbox application separately: either by downloading archive and unpacking to some directory or by using installation package provided for your system. For example, for Debian you may want to install package `nautilus-dropbox`, it will automatically download and install Dropbox application.

How to use
----------

You need Dolphin installed, of course. Then just install dolphin-dropbox-plugin, open Dolphin, go to menu *Settings* → *Configure Dolphin...*, then go to page *Services* and turn on checkbox *Dropbox* in the services list (usually it's enabled by default).

After that you'll get context menu for files and folders in Dropbox directory. Plugin provides basic dropbox services:

* generating link to file from directory Public;
* generating link to any file or folder from Dropbox folder (so called “shareable link”);
* viewing file modifications history;
* and some more less important feautures.

Development notes
=================

To compile project you first need to install kde development packages:

    sudo apt-get install kdelibs5-dev cmake

Compilation:

    mkdir build
    cd build
    cmake -DCMAKE_INSTALL_PREFIX=/usr ..

How to build debian package:

    sh prepare-deb-build.sh _deb

After that you'll find just built package in the directory _deb.


CHANGES.txt
1.4 (2013-Jul-18)

* fixed incorrect context menu for folder (added missing menu item)

1.3 (2013-Jul-07)

* commands in context menu are not guessed now, instead plugin display what dropbox tells to show

1.2 (2012-xx-xx)
----------------

* is dropbox running check added (do not show dropbox context menu items whe dropbox service is not started)
* more intelligent check for context menu availability (using methods provided by Dropbox itsekf)
* russian localization added

1.1 (2012-04-12)
----------------

Complete rewrite.

1.0
---

Initial release, debianized sources downloaded from
http://gitweb.kde.org/scratch/trichard/dolphin-box-plugin.git


prepare-deb-build.sh
#!/bin/bash

if [ "$1" = "" ]; then
    echo "Dest dir required"
    exit 1
fi

if [ ! -e "$1" ]; then
    mkdir "$1"
fi

if [ ! -d "$1" ]; then
    echo "'$1' is not a directory"
    exit 1
fi

DESTDIR=$1
DEFDIR=$(pwd)
PRODNAME=dolphin-dropbox-plugin

echo $DEFDIR

# we need to prepare sources for build
rm -rf $DESTDIR/*

mkdir $DESTDIR/_tmp
cp -r * $DESTDIR/_tmp
cd $DESTDIR/_tmp
rm -rf prepare-deb-build.sh
mv debian ../_deb

# detect program version
cd $DEFDIR
DEB_VERSION=$(head -n1 debian/changelog | sed -e 's/^[^(]*(\([^)]*\)).*/\1/')
UP_VERSION=$(echo $DEB_VERSION | sed -e 's/-.*$//')

# prepare dirs
cd $DESTDIR

BASEDIR="${PRODNAME}-${UP_VERSION}"
mv _tmp "$BASEDIR"
tar czf "${PRODNAME}_${UP_VERSION}.orig.tar.gz" "$BASEDIR"
mv _deb "$BASEDIR"/debian

# ready to build
cd "$BASEDIR"
dpkg-buildpackage -rfakeroot


i18n/extract-messages.sh
#!/bin/sh
# script is taken from http://techbase.kde.org/Development/Tutorials/Localization/i18n_Build_Systems#Extracting_and_merging_messages

BASEDIR="../src/"	# root of translatable sources
PROJECT="dolphin-plugin-dropbox"	# project name
BUGADDR="https://bitbucket.org/cancel/pyrone/issues/new"	# MSGID-Bugs
WDIR=`pwd`		# working dir
 
echo "Preparing rc files"
cd ${BASEDIR}
# we use simple sorting to make sure the lines do not jump around too much from system to system
find . -name '*.rc' -o -name '*.ui' -o -name '*.kcfg' | sort > ${WDIR}/rcfiles.list
xargs --arg-file=${WDIR}/rcfiles.list extractrc > ${WDIR}/rc.cpp
# additional string for KAboutData
echo 'i18nc("NAME OF TRANSLATORS","Your names");' >> ${WDIR}/rc.cpp
echo 'i18nc("EMAIL OF TRANSLATORS","Your emails");' >> ${WDIR}/rc.cpp
cd ${WDIR}
echo "Done preparing rc files"
 
 
echo "Extracting messages"
cd ${BASEDIR}
# see above on sorting
find . -name '*.cpp' -o -name '*.h' -o -name '*.c' | sort > ${WDIR}/infiles.list
echo "rc.cpp" >> ${WDIR}/infiles.list
cd ${WDIR}
xgettext --from-code=UTF-8 -C -kde -ci18n -ki18n:1 -ki18nc:1c,2 -ki18np:1,2 -ki18ncp:1c,2,3 -ktr2i18n:1 \
	-kI18N_NOOP:1 -kI18N_NOOP2:1c,2 -kaliasLocale -kki18n:1 -kki18nc:1c,2 -kki18np:1,2 -kki18ncp:1c,2,3 \
	--msgid-bugs-address="${BUGADDR}" \
	--files-from=infiles.list -D ${BASEDIR} -D ${WDIR} -o ${PROJECT}.pot || { echo "error while calling xgettext. aborting."; exit 1; }
echo "Done extracting messages"
 
 
echo "Merging translations"
catalogs=`find . -name '*.po'`
for cat in $catalogs; do
  echo $cat
  msgmerge -o $cat.new $cat ${PROJECT}.pot
  mv $cat.new $cat
done
echo "Done merging translations"
 
 
echo "Cleaning up"
cd ${WDIR}
rm rcfiles.list
rm infiles.list
rm rc.cpp
echo "Done"


src/dolphin-dropbox-plugin.h
/***************************************************************************
 *   Copyright (C) 2012 by Sergei Stolyarov <sergei.stolyarov@regolit.com> *
 *   Copyright (C) 2010 by Thomas Richard <thomas.richard@proan.be>        *
 *   Copyright (C) 2009-2010 by Peter Penz <peter.penz@gmx.at>             *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA            *
 ***************************************************************************/

#ifndef DolphinDropboxPlugin_H
#define DolphinDropboxPlugin_H

#include <kabstractfileitemactionplugin.h>
#include <QDateTime>
#include <QHash>
#include <QVariantList>
#include <QLocalSocket>
#include <QFileInfo>
#include <QTimer>
#include <QVariant>
#include "QJson/serializer.h"
#include "QJson/parser.h"



//The dropbox protocol info can be found in the dropboxd-protocol file
//it can be found here http://dl.dropbox.com/u/4577542/dropbox/dropboxd-protocol.txt
//Thanks Steffen Schuldenzucker!

/**
 * @brief Dropbox implementation for the KVersionControlPlugin interface.
 */
class DolphinDropboxPlugin : public KAbstractFileItemActionPlugin
{
    Q_OBJECT

public:
    DolphinDropboxPlugin(QObject* parent = 0, const QVariantList & args = QVariantList());
    virtual ~DolphinDropboxPlugin();
    virtual QList<QAction*> actions(const KFileItemListProperties & fileItemInfos, QWidget * parentWidget);

private slots:
    void publicLinkActionHandler();
    void openRemoteFolderActionHandler();

private:
    struct Private;
    Private * p;
    QString sendCommand(QVariant string, QLocalSocket *socket, bool inGuiThread = true);
    QVariant parseJson(QByteArray inp);
};
#endif // DolphinDropboxPlugin_H



src/dolphin-dropbox-plugin.cpp
/***************************************************************************
 *   Copyright (C) 2012 by Sergei Stolyarov <sergei@regolit.com>           *
 *   Copyright (C) 2010 by Thomas Richard <thomas.richard@proan.be>        *
 *   Copyright (C) 2009-2010 by Peter Penz <peter.penz@gmx.at>             *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA            *
 ***************************************************************************/
 

#include <kaction.h>
#include <kactionmenu.h>
#include <kdemacros.h>
#include <kfileitem.h>
#include <klocalizedstring.h>
#include <kfileitemlistproperties.h>
#include <QDir>
#include <QFileInfo>
#include <QString>
#include <QStringList>
#include <QLocalSocket>
#include <QApplication>
#include <QClipboard>
#include <QtDBus>
#include <QDesktopServices>

#include <KDE/KPluginFactory>
#include <KDE/KPluginLoader>

#include <QtDebug>

#include "dolphin-dropbox-plugin.h"

#define SOCKET_FILE ".2safe/control.sock"

K_PLUGIN_FACTORY(DolphinDropboxPluginFactory, registerPlugin<DolphinDropboxPlugin>();)
K_EXPORT_PLUGIN(DolphinDropboxPluginFactory("dolphin-dropbox-plugin"))

struct DolphinDropboxPlugin::Private
{
    QString contextFilePath;
    QLocalSocket controlSocket;
    QString controlSocketPath;
};


DolphinDropboxPlugin::DolphinDropboxPlugin(QObject* parent, const QVariantList & args):
    KAbstractFileItemActionPlugin(parent)
{
    Q_UNUSED(args);
    p = new Private;
    p->controlSocketPath = (QDir::homePath() + QDir::separator() + QString(SOCKET_FILE));
}

DolphinDropboxPlugin::~DolphinDropboxPlugin()
{
    delete p;
}

QList<QAction*> DolphinDropboxPlugin::actions(const KFileItemListProperties & fileItemInfos, QWidget * parentWidget)
{
    Q_UNUSED(parentWidget);
    QList<QAction*> actions;

    if (fileItemInfos.items().count() != 1) {
        // do not display menu on empty or multiple selection
        return actions;
    }

    KFileItem item = fileItemInfos.items().first();
    QString itemPath = item.localPath();
    QFileInfo itemFileInfo = QFileInfo(itemPath);

    if (!item.isDir() && !item.isFile() && !item.isLocalFile()) {
        // file is not dir and local file
        return actions;
    }

    if (!itemFileInfo.filePath().contains("/home/xlab/2safe/", Qt::CaseInsensitive) &&
         !itemFileInfo.filePath().contains("/home/xlab/Blog/", Qt::CaseInsensitive)) {
        // file is not in supported dirs
        return actions;
    }

    // everything looks fine so populate the menu
    KActionMenu * menuAction = new KActionMenu(this);
    menuAction->setText("2Safe Plugin");
    menuAction->setIcon(QIcon("/home/xlab/Coding/2safe-dolphin/2safe.png"));
    actions << menuAction;

    p->contextFilePath = itemPath;

    QAction * publicLinkAction = new KAction(this);
    publicLinkAction->setIcon(KIcon("klipper"));
    publicLinkAction->setText(i18nc("@item:inmenu", "Get public link"));
    connect(publicLinkAction, SIGNAL(triggered()),
            this, SLOT(publicLinkActionHandler()));

    QAction * openRemoteFolderAction = new KAction(this);
    openRemoteFolderAction->setIcon(KIcon("konqueror"));
    openRemoteFolderAction->setText(i18nc("@item:inmenu", "Open in remote folder"));
    connect(openRemoteFolderAction, SIGNAL(triggered()),
            this, SLOT(openRemoteFolderActionHandler()));

    menuAction->addAction(publicLinkAction);
    menuAction->addAction(openRemoteFolderAction);

    return actions;
}

//Re-entrant
QString DolphinDropboxPlugin::sendCommand(QVariant command, QLocalSocket *socket, bool inGuiThread)
{
    //If we're in the GuiThread we can't afford to wait long
    //in the worker thread we have bigger margins
    int waitTime = 2000;
    if(inGuiThread)
        waitTime = 1000;

    if(!socket->isOpen()) {
        socket->connectToServer(p->controlSocketPath);
        if(!socket->waitForConnected(waitTime))
            return QString();
    }

    QJson::Serializer serializer;

    QByteArray _jcommand = serializer.serialize(command);
    socket->write(_jcommand);
    socket->flush();

    bool ready = false;
    QString reply;

    if(!socket->waitForReadyRead(waitTime))    {
        socket->close();
        return QString();
    }

    reply.append(socket->readAll());
    return reply;
}

QVariant DolphinDropboxPlugin::parseJson(QByteArray inp){
    QJson::Parser parser;
    bool ok;
    QVariant out =parser.parse(inp,&ok);
    if (!ok)
        return QVariant();
    return out;
}

void DolphinDropboxPlugin::publicLinkActionHandler(){
    QVariantMap command;
    QVariantMap args;
    args.insert("file", p->contextFilePath);
    command.insert("type","action");
    command.insert("verb","get_public_link");
    command.insert("args", args);
    QString link = sendCommand(command, &(p->controlSocket));
    if(!link.isEmpty()) {
        //QDesktopServices::openUrl(QUrl(link));
        QApplication::clipboard()->setText(link);
    }
}

void DolphinDropboxPlugin::openRemoteFolderActionHandler(){
    QVariantMap command;
    QVariantMap args;
    args.insert("file", p->contextFilePath);
    command.insert("type","action");
    command.insert("verb","open_in_browser");
    command.insert("args", args);
    QString link = sendCommand(command, &(p->controlSocket));
    if(!link.isEmpty()){
        QDesktopServices::openUrl(QUrl(link));
    }
}